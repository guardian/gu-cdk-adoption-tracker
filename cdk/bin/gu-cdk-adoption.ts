// This file was autogenerated when creating gu-cdk-adoption.ts using @guardian/cdk-cli
// It is a starting point for migration to CDK *only*. Please check the output carefully before deploying

import "source-map-support/register";
import { App } from "@aws-cdk/core";
import { GuCdkAdoption } from "../lib/gu-cdk-adoption";

const app = new App();
new GuCdkAdoption(app, "GuCdkAdoption", {
  app: "gu-cdk-adoption",
  stack: "deploy",
});
