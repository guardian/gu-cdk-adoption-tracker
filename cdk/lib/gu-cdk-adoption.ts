// This file was autogenerated using @guardian/cdk-cli

import {Schedule} from "@aws-cdk/aws-events";
import {PolicyStatement} from "@aws-cdk/aws-iam";
import {Runtime} from "@aws-cdk/aws-lambda";
import type {App} from "@aws-cdk/core";
import {Duration} from "@aws-cdk/core";
import {GuScheduledLambda} from "@guardian/cdk";
import type {GuStackProps} from "@guardian/cdk/lib/constructs/core";
import {GuDistributionBucketParameter, GuStack, GuStringParameter,} from "@guardian/cdk/lib/constructs/core";
import {GuVpc} from "@guardian/cdk/lib/constructs/ec2";
import {TreatMissingData} from "@aws-cdk/aws-cloudwatch";

export class GuCdkAdoption extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const bucketName = new GuDistributionBucketParameter(this, "dist-bucket");
    const apiToken = new GuStringParameter(this, "api-token", {
      description: "GitHub API token with repo permissions",
      fromSSM: true,
    });
    const prismUrl = new GuStringParameter(this, "prism-url", {
      fromSSM: true,
    });

    const lambda = new GuScheduledLambda(this, "scheduled-lambda", {
      code: {
        bucket: bucketName.valueAsString,
        key: `${this.stack}/${this.stage}/${this.app}/${this.app}.zip`,
      },
      environment: {
        API_TOKEN: apiToken.valueAsString,
        PRISM_URL: prismUrl.valueAsString
      },
      functionName: `${this.app}-${this.stage}`,
      monitoringConfiguration: {
        toleratedErrorPercentage: 50,
        numberOfFiveMinutePeriodsToEvaluate: 3,
        treatMissingData: TreatMissingData.NOT_BREACHING, //TODO fix this at construct level
        snsTopicName: "devx-alerts",
      },
      schedule: Schedule.rate(Duration.days(1)),
      runtime: Runtime.NODEJS_14_X,
      handler: "lib/handler.handler",
      vpc: GuVpc.fromIdParameter(this, "vpc"),
      vpcSubnets: { subnets: GuVpc.subnetsfromParameter(this) },
    });

    lambda.addToRolePolicy(
      new PolicyStatement({
        actions: ["cloudwatch:PutMetricData"],
        resources: ["*"],
      })
    );
  }
}
