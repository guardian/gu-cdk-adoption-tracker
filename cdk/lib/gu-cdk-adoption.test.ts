// This file was autogenerated using @guardian/cdk-cli

import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { GuCdkAdoption } from "./gu-cdk-adoption";

describe("The GuCdkAdoption stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new GuCdkAdoption(app, "gu-cdk-adoption", {app: "gu-cdk-adoption", stack: "deploy" });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
