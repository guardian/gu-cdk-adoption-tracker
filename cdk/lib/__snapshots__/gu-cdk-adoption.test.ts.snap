// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuCdkAdoption stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "PrivateSubnets": Object {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
    "VpcId": Object {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "apitoken": Object {
      "Description": "GitHub API token with repo permissions",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "distbucket": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "prismurl": Object {
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ErrorPercentageAlarmForLambda54F48384": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":devx-alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "scheduledlambdaDBC85EB4",
              },
              " exceeded 50% error rate",
            ],
          ],
        },
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "High error % from ",
              Object {
                "Ref": "scheduledlambdaDBC85EB4",
              },
              " lambda in ",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "Error % of ",
                  Object {
                    "Ref": "scheduledlambdaDBC85EB4",
                  },
                ],
              ],
            },
          },
          Object {
            "Id": "m1",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "scheduledlambdaDBC85EB4",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m2",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "scheduledlambdaDBC85EB4",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 50,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "scheduledlambdaDBC85EB4": Object {
      "DependsOn": Array [
        "scheduledlambdaServiceRoleDefaultPolicy950F82AB",
        "scheduledlambdaServiceRole3CE873B7",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "distbucket",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                "deploy/",
                Object {
                  "Ref": "Stage",
                },
                "/gu-cdk-adoption/gu-cdk-adoption.zip",
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "API_TOKEN": Object {
              "Ref": "apitoken",
            },
            "PRISM_URL": Object {
              "Ref": "prismurl",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "gu-cdk-adoption-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "Handler": "lib/handler.handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "scheduledlambdaServiceRole3CE873B7",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "gu-cdk-adoption",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "0.39.1",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Timeout": 30,
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "scheduledlambdaSecurityGroup077C1AF2",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Object {
            "Ref": "PrivateSubnets",
          },
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "scheduledlambdaSecurityGroup077C1AF2": Object {
      "Properties": Object {
        "GroupDescription": "Automatic security group for Lambda Function gucdkadoptionscheduledlambdaFA4459E7",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "gu-cdk-adoption",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "0.39.1",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "scheduledlambdaServiceRole3CE873B7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "gu-cdk-adoption",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "0.39.1",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "scheduledlambdaServiceRoleDefaultPolicy950F82AB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "distbucket",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "distbucket",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "scheduledlambdaServiceRoleDefaultPolicy950F82AB",
        "Roles": Array [
          Object {
            "Ref": "scheduledlambdaServiceRole3CE873B7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "scheduledlambdascheduledlambdarate1day032EF8A05": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 day)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "scheduledlambdaDBC85EB4",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "scheduledlambdascheduledlambdarate1day0AllowEventRulegucdkadoptionscheduledlambdascheduledlambdarate1day020E381A0627B44DE": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "scheduledlambdaDBC85EB4",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "scheduledlambdascheduledlambdarate1day032EF8A05",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
